reflekt-plugin/src/main/kotlin/io/reflekt/plugin/analysis/processor/Util.kt

```diff
@@ -10,14 +12,67 @@ internal val KtElement.isPublicObject: Boolean
 internal val KtElement.isPublicFunction: Boolean
     get() = this is KtNamedFunction && this.isPublic
 
+internal val KtElement.isTopLevelPublicFunction: Boolean
+    get() = isPublicFunction && this is KtNamedFunction && this.isTopLevel
+
 internal val KtElement.isMainFunction: Boolean
-    get() = this.isPublicFunction && (this as KtNamedFunction).nameIdentifier?.let { it.text == "main" } ?: false
+    get() = isPublicFunction && (this as KtNamedFunction).nameIdentifier?.let { it.text == "main" } ?: false
 
 internal val KtElement.isPublicNotAbstractClass: Boolean
     get() = this is KtClass && this.isPublic && !this.isAbstract()
 
 typealias FileID = String
 
+internal fun getFullName(packageFqName: FqName, name: String? = null) : FileID {
+    val postfix = name?.let{ ".${name}" } ?: ""
+    return "${packageFqName.asString()}$postfix"
+}
+
 // TODO: is it enough to identify a file?
 internal val KtFile.fullName: FileID
-    get() = "${this.packageFqName.asString()}.${this.name}"
+    get() = getFullName(this.packageFqName, this.name)
+
+internal val DeclarationDescriptor.isPublicNotAbstractClass: Boolean
+    get() = this is ClassDescriptor && this.isClass && !this.isAbstractClass && this.isPublic
+
+internal val DeclarationDescriptor.isPublicObject: Boolean
+    get() = this is ClassDescriptor && this.isObject && this.isPublic
+
+internal val DeclarationDescriptor.isPublicTopLevelFunction: Boolean
+    get() = this is FunctionDescriptor && this.isPublic && this.isTopLevelFunction
+
+internal val DeclarationDescriptor.isMainFunction: Boolean
+    get() = isPublicTopLevelFunction && (this as FunctionDescriptor).name.asString() == "main"
+
+internal val FunctionDescriptor.isTopLevelFunction: Boolean
+    get() = this.isTopLevelInPackage()
+
+internal val FunctionDescriptor.isPublic: Boolean
+    get() = this.visibility.isPublicAPI
+
+internal val ClassDescriptor.isAbstractClass: Boolean
+    get() = this.modality == Modality.ABSTRACT
+
+internal val ClassDescriptor.isPublic: Boolean
+    get() = this.visibility.isPublicAPI
+
+internal val ClassDescriptor.isClass: Boolean
+    get() = this.kind != ClassKind.OBJECT && this.kind != ClassKind.ENUM_ENTRY
+
+internal val ClassDescriptor.isObject: Boolean
+    get() = this.kind == ClassKind.OBJECT
+
+internal fun <K, V : MutableSet<K>> getInvokesGroupedByFiles(fileToInvokes: HashMap<FileID, V>) =
+    groupFilesByInvokes(fileToInvokes).keys.flatten().toMutableSet()
+
+// To avoid repeated checks for belonging invokes in different files,
+// we will group files by invokes and process each of them once
+// MutableSet<*> here is ClassOrObjectInvokes = MutableSet<SupertypesToAnnotations>
```

Comment:
which `MutableSet<*>` are you talking about?

Reply:
I added a bound, thank you!